optimism/src/keccak/constraints.rs

```diff
@@ -26,8 +30,38 @@ impl<Fp: Field> Constraints for KeccakEnv<Fp> {
     }
 
     fn constraints(&mut self) {
-        todo!();
         // CORRECTNESS OF FLAGS
+        {
+            // TODO: remove redundancy if any
+
+            // Booleanity of sponge flags
+            {
+                // Absorb is either true or false
+                self.constrain(Self::boolean(self.absorb()));
+                // Squeeze is either true or false
+                self.constrain(Self::boolean(self.squeeze()));
+                // Root is either true or false
+                self.constrain(Self::boolean(self.root()));
+                // Pad is either true or false
+                self.constrain(Self::boolean(self.pad()));
+            }
+            // Mutually exclusiveness of flags
+            {
+                // Squeeze and Root are not both true
+                self.constrain(Self::either_false(self.squeeze(), self.root()));
```

Comment:
What's the expected output of `either_false`?
If at least one of the parameters is `false` (i.e. `0`), return `true` (i.e. `1`)?

Reply:
if at least one of them is false, then the constrain is satisfied -> the constrain is not satisfied if both are true -> you cannot have an absorb root in a squeeze sponge